ocaml/dune
# tui: add more interactions

We add some more features to the `--display tui` mode.

Waiting for #7663 

Features:
- Messages to the user a separated with a rule and include a numbering together with total number of messages.
- Horizontal and vertical scroll bars which can be controlled using the arrow keys, `j` and `k` or by interacting with the mouse including scrolling and clicking. (Horizontal scrolling is done with Meta (alt) + Scroll.
- Help screen activated with `h` or `?` explaining what keys do what.
- Debug features:
  - Pressing 'd' will output some debug info about the state of the UI.
  - Pressing Meta + `l` will make dune output a lorem ipsum, allowing to quickly test any issues with scrolling. (I want to keep this in as it is harmless and out of the way).

<img src="https://user-images.githubusercontent.com/8614547/235651225-e4aab871-c578-47e6-96ff-f9446c877d87.png" width="500">
<img src="https://user-images.githubusercontent.com/8614547/235651220-dc99467a-1467-4736-b982-08d4c42dd158.png" width="500">

#### How to test:
Just run
```console
./dune.exe build @install
_build/install/bin/dune build @check -w --display tui
```

#### Compatibility
- Linux
  - [x] Alacritty, works fine, this is in the screenshots
  - [x] xterm, works but meta + l is sometimes ignored
  - [x] gnome console, works fine
  - [ ] cmd
  - [ ] wsl

#### Issues
- [ ] Exceptions are not printed when in tui mode. We need to have a fallback display. For example, running two watch mode tui's will cause the second one to crash silently and return to the prompt since Dune will complain about the _build lock. We should probably add a way internal exceptions can be displayed or printed nicely.
- [x] Needs to look good on light terminals, the colors used are somewhat ad-hoc. We need to make them more configurable and then adapt based on terminal colors chosen.
